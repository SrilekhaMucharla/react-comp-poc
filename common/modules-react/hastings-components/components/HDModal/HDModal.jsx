/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/no-static-element-interactions */
import React, { useEffect, useState } from 'react';
import { PropTypes } from 'prop-types';
import { createPortal } from 'react-dom';
import { Row, Col, Container } from 'react-bootstrap';
import classNames from 'classnames';
import HDButtonRefactor from '../HDButtonRefactor/HDButtonRefactor';
import HDLabelRefactor from '../HDLabelRefactor/HDLabelRefactor';
import crossCircleBlack from '../../../../../applications/quote-and-buy/src/assets/images/icons/cross_circle_black.svg';

const HDModal = (props) => {
    const {
        show,
        headerText,
        cancelLabel,
        confirmLabel,
        onCancel,
        onConfirm,
        onClose,
        onBeforeOpen,
        children,
        customStyle,
        className,
        hideCancelButton,
        hideConfirmButton,
        hideFooter,
        hideClose,
        disableConfirm
    } = props;

    const [isStacked, setIsStacked] = useState(false);
    const [customStyles, setCustomStyles] = useState([]);
    const [isVisibible, setIsVisible] = useState(false);
    const [showContent, setShowContent] = useState(false);

    const divToRender = document.getElementById('portal-root');

    useEffect(() => {
        if (show && onBeforeOpen) {
            onBeforeOpen();
        }
    }, [show]);

    // clean up on unmount
    useEffect(() => {
        return () => {
            if (document.body.classList.contains('hd-modal-open')) {
                document.body.classList.remove('hd-modal-open');
            }
        };
    }, []);


    useEffect(() => {
        setCustomStyles(customStyle.split(' '));
    }, [customStyle]);

    const fullWidthFooterBtns = hideCancelButton || hideCancelButton || customStyles.includes('footer-btns-w-100');
    const revBtnOrder = customStyles.includes('rev-button-order');

    const footer = (
        <Container fluid className="p-0">
            {customStyles.includes('wide')
                ? (
                    <Row
                        className={classNames(
                            'modal__footer',
                            { 'footer-btns-w-100': fullWidthFooterBtns },
                            { 'flex-column-reverse': revBtnOrder },
                            { 'flex-sm-row-reverse': revBtnOrder && !fullWidthFooterBtns }
                        )}
                    >
                        {!hideConfirmButton && (
                            <Col
                                xs={12}
                                sm={!fullWidthFooterBtns ? 6 : undefined}
                                className={classNames(
                                    { 'modal-btn-pr-small': !fullWidthFooterBtns },
                                    { 'pt-2': !hideCancelButton && revBtnOrder },
                                    { 'pt-sm-0': !hideCancelButton && !fullWidthFooterBtns && revBtnOrder }
                                )}
                            >
                                <HDButtonRefactor
                                    className="btn-primary btn-block theme-white"
                                    label={confirmLabel}
                                    onClick={onConfirm}
                                    data-testid="confirm-button"
                                    variant="primary"
                                    disabled={disableConfirm} />

                            </Col>
                        )}
                        {!hideCancelButton && (
                            <Col
                                xs={12}
                                sm={!fullWidthFooterBtns ? 6 : undefined}
                                className={classNames(
                                    { 'modal-btn-pl-small': !fullWidthFooterBtns },
                                    { 'pt-3': !hideConfirmButton && !revBtnOrder },
                                    { 'pt-sm-0': !hideConfirmButton && !fullWidthFooterBtns && !revBtnOrder }
                                )}
                            >
                                <HDButtonRefactor
                                    className="btn-secondary btn-block theme-white"
                                    label={cancelLabel}
                                    onClick={onCancel}
                                    data-testid="cancel-button"
                                    variant="secondary" />
                            </Col>
                        )}
                    </Row>
                )
                : (
                    <Row
                        className={classNames(
                            'modal__footer',
                            { 'footer-btns-w-100': fullWidthFooterBtns },
                            { 'flex-column-reverse': revBtnOrder },
                            { 'flex-sm-row-reverse': revBtnOrder && !fullWidthFooterBtns }
                        )}
                    >
                        {!hideCancelButton && (
                            <Col
                                xs={12}
                                sm={!fullWidthFooterBtns ? 6 : undefined}
                                className={classNames(
                                    { 'modal-btn-pr-small': !fullWidthFooterBtns },
                                    { 'pt-3': !hideConfirmButton && revBtnOrder },
                                    { 'pt-sm-0': !hideConfirmButton && !fullWidthFooterBtns && revBtnOrder }
                                )}
                            >
                                <HDButtonRefactor
                                    className="btn-secondary btn-block theme-white"
                                    label={cancelLabel}
                                    onClick={onCancel}
                                    data-testid="cancel-button" />

                            </Col>
                        )}
                        {!hideConfirmButton && (
                            <Col
                                xs={12}
                                sm={!fullWidthFooterBtns ? 6 : undefined}
                                className={classNames(
                                    { 'modal-btn-pl-small': !fullWidthFooterBtns },
                                    { 'pt-2': !hideCancelButton && !revBtnOrder },
                                    { 'pt-sm-0': !hideCancelButton && !fullWidthFooterBtns }
                                )}
                            >
                                <HDButtonRefactor
                                    className="btn-primary btn-block theme-white"
                                    label={confirmLabel}
                                    onClick={onConfirm}
                                    data-testid="confirm-button" />

                            </Col>
                        )}
                    </Row>
                )}
        </Container>
    );

    useEffect(() => {
        if (show) {
            setShowContent(true);
        } else {
            setIsVisible(false);
            if (customStyle === 'wide') {
                setTimeout(() => {
                    setShowContent(false);
                }, 500);
            } else {
                setShowContent(false);
            }
        }
    }, [show]);

    useEffect(() => {
        if (showContent) {
            if (document.body.classList.contains('hd-modal-open')) {
                setIsStacked(true);
            } else {
                document.body.classList.add('hd-modal-open');
            }
            setIsVisible(true);
        } else if (!isStacked) {
            document.body.classList.remove('hd-modal-open');
        } else {
            setIsStacked(false);
        }
    }, [showContent]);

    const modalClassNames = classNames(
        `hastings-modal theme-white ${className} ${customStyle}`,
        { narrow: customStyle !== 'wide' },
    );

    const contentClassNames = classNames(
        'hastings-modal-content',
        { 'is-visible': isVisibible }
    );

    const dialogBox = () => {
        return (
            <>
                {showContent
                    ? (
                        <div className={modalClassNames}>
                            <div className={contentClassNames}>
                                {(headerText || !hideClose) && (
                                    <div className="hd-modal-header">
                                        {!hideClose && (
                                            <span data-testid="close-button" onClick={onClose}>
                                                <img src={crossCircleBlack} alt="cross_circle_black_img" />
                                            </span>
                                        )}
                                        {headerText && (
                                            <HDLabelRefactor Tag="h2" text={headerText} />
                                        )}
                                    </div>
                                )}
                                <div className="hd-modal-body">
                                    {children}
                                </div>
                                <div className="hd-modal-footer">
                                    {!hideFooter && footer}
                                </div>
                            </div>
                        </div>
                    ) : ''}
            </>
        );
    };

    return (
        createPortal(dialogBox(), divToRender)
    );
};

HDModal.propTypes = {
    show: PropTypes.bool.isRequired,
    headerText: PropTypes.string,
    cancelLabel: PropTypes.string,
    confirmLabel: PropTypes.string,
    onCancel: PropTypes.func,
    onConfirm: PropTypes.func.isRequired,
    onClose: PropTypes.func,
    children: PropTypes.node.isRequired,
    customStyle: PropTypes.string,
    className: PropTypes.string,
    hideCancelButton: PropTypes.bool,
    hideConfirmButton: PropTypes.bool,
    hideFooter: PropTypes.bool,
    hideClose: PropTypes.bool,
    onBeforeOpen: PropTypes.func
};

HDModal.defaultProps = {
    headerText: '',
    cancelLabel: 'Cancel',
    confirmLabel: 'Confirm',
    onCancel: () => { },
    onClose: () => { },
    onBeforeOpen: () => {},
    customStyle: '',
    className: '',
    hideFooter: false,
    hideCancelButton: false,
    hideConfirmButton: false,
    hideClose: false,
};

export default HDModal;
