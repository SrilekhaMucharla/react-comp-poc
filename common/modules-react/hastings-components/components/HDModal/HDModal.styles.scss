.hastings-modal {
    position: fixed;
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
    width: 100%;
    height: 100%;
    background: $loader-bg;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    z-index: 1000000;

    .hastings-modal-content {
        width: 100%;
        max-width: 371px;
        margin: 0 20px;
        position: relative;
        background: $white-fill;
        border-radius: 5px;
        padding: $lg-value;
        @include media-breakpoint-up(md) {
            padding: 30px 43px;
        }
        h1,
        h2,
        h3,
        h4,
        h5 {
            color: $mono-midnight;
        }

        h5 {
            margin-top: $lg-value;
            margin-bottom: $md-value;
            color: $midnight-nordic-noir;
        }

        p {
            color: $mono-midnight;
        }

        hr {
            height: 0.5px;
        }
    }

    .hd-modal-header {
        text-align: left;

        > span {
            display: block;
            text-align: right;

            > img {
                height: 36px;
                width: 37px;
                cursor: pointer;
            }
        }
    }

    .hd-modal-body {
        font-family: $font-face-generic-regular;
        font-style: normal;
        font-weight: normal;
        font-size: $font-size-md;
        line-height: 24px;
        color: $app-theme;

        .btn-group-toggle .btn {
            @extend .btn--light;
        }

        .drop-down-list > div > div > div {
            font-weight: 600;
        }
    }

    .hd-modal-footer {
        display: flex;
        margin-top: 20px;

        .flex-column-reverse,
        .footer-btns-w-100 {
            button {
                height: 50px;
            }
        }

        @include media-breakpoint-up(sm) {
            margin-top: 30px;

            .row:not(.flex-column-reverse) {
                .modal-btn-pl-small {
                    padding-left: 6px;
                }

                .modal-btn-pr-small {
                    padding-right: 6px;
                }
            }

            .flex-column-reverse {
                .modal-btn-pl-small {
                    padding-right: 6px;
                }

                .modal-btn-pr-small {
                    padding-left: 6px;
                }
            }
        }
    }

    &.narrow {
        .hd-modal-header {
            h2 {
                font-size: $hd-font-size-h2-mobile;
                margin-top: 0;
                margin-bottom: 9px;
            }
        }
        .hd-modal-footer {
            margin-top: 27px;
        }
    }

    &.wide {
        justify-content: flex-end;
        .hastings-modal-content {
            position: absolute;
            left: 100%;
            margin: 0;
            padding: 60px 20px 40px 20px;
            width: 100%;
            max-width: 100%;
            height: 100%;
            border-radius: 0;
            overflow: auto;

            @include media-breakpoint-up(md) {
                padding: 45px 76px 40px 67px;
                width: 50%;
                min-width: 559px;
            }
            @include media-breakpoint-up(xl) {
                width: 30%;
            }

            .hd-modal-header {
                > span {
                    position: absolute;
                    z-index: 1;
                    top: 16px;
                    right: 20px;
                    @include media-breakpoint-up(md) {
                        top: 35px;
                        right: 32px;
                    }
                    > img {
                        height: 28px;
                        width: 28px;
                        @include media-breakpoint-up(md) {
                            height: 36px;
                            width: 37px;
                        }
                    }
                }
            }
        }
    }

    .modal__footer.flex-row-reverse {
        button {
            font-weight: 600;
        }
    }
}

//TODO: Add non-wide style

//TODO: review implementation
.hastings-modal.ancillary-popup {
    .hastings-modal-content {
        align-items: center;
        justify-content: center;
        padding: 20px;
        min-height: 0;
        width: 375px;
        background-color: $black-transparent-25;

        .hd-modal-header,
        .hd-modal-footer {
            display: none;
        }

        .hd-modal-body {
            background-color: $app-theme;
            margin-bottom: 0;
            margin-left: 0;
            margin-right: 0;

            .popup-card {
                padding: 10px;
                display: flex;

                .popup-message {
                    font-size: $font-size-regular;
                    text-align: left;
                    color: $white-color;
                }

                .popup-icon {
                    width: 10px;
                    height: 10px;
                    background-position: center;
                    background-repeat: no-repeat;
                    background-size: contain;
                    padding: 12px;
                    border-radius: 50%;
                    margin-right: 12px;
                }

                .popup-icon.popup-icon-added {
                    background-image: url(../../../../../applications/quote-and-buy/src/assets/images/wizard-images/hastings-icons/icons/Tick.svg);
                }

                .popup-icon.popup-icon-removed {
                    background-color: $alert-color;
                    background-image: url(../../../../../applications/quote-and-buy/src/assets/images/wizard-images/hastings-icons/icons/Cross.svg);
                }
            }
        }
    }
}
