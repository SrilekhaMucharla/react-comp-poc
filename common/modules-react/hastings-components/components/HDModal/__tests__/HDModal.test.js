import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import HDModal from '../HDModal';

Enzyme.configure({ adapter: new Adapter() });

describe('<HDModal/>', () => {
    it('should render component', () => {
        const onCloseMock = jest.fn();
        const onConfirmMock = jest.fn();
        const onCancelMock = jest.fn();
        const component = (
            <div id="portal-modal">
                <HDModal show onClose={onCloseMock} onConfirm={onConfirmMock} onCancel={onCancelMock}>
                    <div>Test</div>
                </HDModal>
            </div>
        );
        const wrapper = shallow(component);
        expect(wrapper).toMatchSnapshot();
    });
    // Header text
    it('should render component with header text', () => {
        const onCloseMock = jest.fn();
        const onConfirmMock = jest.fn();
        const onCancelMock = jest.fn();
        const component = (
            <div id="portal-modal">
                <HDModal headerText="Heading" show onClose={onCloseMock} onConfirm={onConfirmMock} onCancel={onCancelMock}>
                    <div>Test</div>
                </HDModal>
            </div>
        );
        const wrapper = shallow(component);
        expect(wrapper).toMatchSnapshot();
    });

    it('should render component with header text and confirm', () => {
        const onCloseMock = jest.fn();
        const onConfirmMock = jest.fn();
        const onCancelMock = jest.fn();
        const component = (
            <div id="portal-modal">
                <HDModal headerText="Heading" confirmLabel="Yes" show onClose={onCloseMock} onConfirm={onConfirmMock} onCancel={onCancelMock}>
                    <div>Test</div>
                </HDModal>
            </div>
        );
        const wrapper = shallow(component);
        expect(wrapper).toMatchSnapshot();
    });
    it('should render component with header text confirm and cancel', () => {
        const onCloseMock = jest.fn();
        const onConfirmMock = jest.fn();
        const onCancelMock = jest.fn();
        const component = (
            <div id="portal-modal">
                <HDModal headerText="Heading" confirmLabel="Yes" cancelLabel="Go" show onClose={onCloseMock} onConfirm={onConfirmMock} onCancel={onCancelMock}>
                    <div>Test</div>
                </HDModal>
            </div>
        );
        const wrapper = shallow(component);
        expect(wrapper).toMatchSnapshot();
    });
});
