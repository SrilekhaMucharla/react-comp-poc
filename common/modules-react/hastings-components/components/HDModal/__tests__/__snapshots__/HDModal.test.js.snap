// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HDModal/> should render component 1`] = `
<div
  id="portal-modal"
>
  <HDModal
    cancelLabel="Cancel"
    className=""
    confirmLabel="Confirm"
    customStyle=""
    headerText=""
    hideCancelButton={false}
    hideClose={false}
    hideConfirmButton={false}
    hideFooter={false}
    onBeforeOpen={[Function]}
    onCancel={[MockFunction]}
    onClose={[MockFunction]}
    onConfirm={[MockFunction]}
    show={true}
  >
    <div>
      Test
    </div>
  </HDModal>
</div>
`;

exports[`<HDModal/> should render component with header text 1`] = `
<div
  id="portal-modal"
>
  <HDModal
    cancelLabel="Cancel"
    className=""
    confirmLabel="Confirm"
    customStyle=""
    headerText="Heading"
    hideCancelButton={false}
    hideClose={false}
    hideConfirmButton={false}
    hideFooter={false}
    onBeforeOpen={[Function]}
    onCancel={[MockFunction]}
    onClose={[MockFunction]}
    onConfirm={[MockFunction]}
    show={true}
  >
    <div>
      Test
    </div>
  </HDModal>
</div>
`;

exports[`<HDModal/> should render component with header text and confirm 1`] = `
<div
  id="portal-modal"
>
  <HDModal
    cancelLabel="Cancel"
    className=""
    confirmLabel="Yes"
    customStyle=""
    headerText="Heading"
    hideCancelButton={false}
    hideClose={false}
    hideConfirmButton={false}
    hideFooter={false}
    onBeforeOpen={[Function]}
    onCancel={[MockFunction]}
    onClose={[MockFunction]}
    onConfirm={[MockFunction]}
    show={true}
  >
    <div>
      Test
    </div>
  </HDModal>
</div>
`;

exports[`<HDModal/> should render component with header text confirm and cancel 1`] = `
<div
  id="portal-modal"
>
  <HDModal
    cancelLabel="Go"
    className=""
    confirmLabel="Yes"
    customStyle=""
    headerText="Heading"
    hideCancelButton={false}
    hideClose={false}
    hideConfirmButton={false}
    hideFooter={false}
    onBeforeOpen={[Function]}
    onCancel={[MockFunction]}
    onClose={[MockFunction]}
    onConfirm={[MockFunction]}
    show={true}
  >
    <div>
      Test
    </div>
  </HDModal>
</div>
`;
